torch/optim/adam.py

```diff
@@ -108,6 +118,14 @@ def _init_group(
                     max_exp_avg_sqs.append(state['max_exp_avg_sq'])
                 if group['differentiable'] and state['step'].requires_grad:
                     raise RuntimeError('`requires_grad` is not supported for `step` in differentiable mode')
+
+                # Host lr as a tensor when fused to allow a dynamic learning rate
+                # but temporarily disallow tensor lrs for other implementations to avoid host2device syncs
+                if group['fused'] and not torch.is_tensor(group['lr']):
+                    group['lr'] = torch.tensor(group['lr'], dtype=torch.float, device=p.device)
+                elif not group['fused'] and torch.is_tensor(group['lr']):
+                    group['lr'] = group['lr'].item()
```

Comment:
We do this because our fop-loop and foreach code actually raises an error when given a tensor()? Do we test this?

Reply:
We do this because it's better to set the lr correctly globally when one calls step through the optimizer. The for-loop and foreach code should never raise user-level errors though because we coerce the types in init_group and adam() which are the exposed endpoints.

But yes, agreed, I should add tests for passing tensors as lr for the forloop and foreach impls too.

Reply:
So today this `.item()` call happens anyways (explicitly or implicitly) during the step() right?

Reply:
only if the user calls adam directly  as a functional api. if the user inits optimizers through the constructor, the .item is only called during the first step in init_group.

we could have our forloop and foreach case error instead if we want to be blatant about the behavior

Reply:
The .item only happens in PT2 so tehre's no need to blatantly coerce in eager